### PR TITLE
Databricks SDK based models artifact repo

### PR DESCRIPTION
**Pull request by shichengzhou-db**
### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Context:
A small subset of Databricks SEG(workspaces with Internet access turned off) customers will be blocked from using MLFlow due to the fact that mlflow client talks to S3/Azure for model registry. This PR will unblock those customers by allowing model artifacts download/upload through Databricks SDK which internally uses Presigned URL / Files API.

- If MLFLOW_USE_DATABRICKS_SDK_MODEL_ARTIFACTS_REPO_FOR_UC is set to true, use the new `DatabricksSDKModelsArtifactRepository` to register and download models to and from UC.
- `DatabricksSDKModelsArtifactRepository` implements `CloudArtifactRepository` to list artifacts, download/upload artifacts to cloud using the [Databricks SDK Files interface](https://databricks-sdk-py.readthedocs.io/en/latest/workspace/files/files.html).

### How is this PR tested?
- Unit Tests
- Download / Upload models to databricks workspace works

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/store/artifact/unity_catalog_models_artifact_repo.py
Comment (harupy) [inline mlflow/store/artifact/unity_catalog_models_artifact_repo.py:131]: `eprint` should be used for an error or a warning. Is this an error or a warning?

File: tests/store/artifact/test_databricks_sdk_models_artifact_repo.py
Comment (harupy) [inline tests/store/artifact/test_databricks_sdk_models_artifact_repo.py:1]: can we test the MLFLOW_USE_DATABRICKS_SDK_MODEL_ARTIFACTS_REPO_FOR_UC env var works correctly?

File: mlflow/store/artifact/unity_catalog_models_artifact_repo.py
Comment (shichengzhou-db) [reply mlflow/store/artifact/unity_catalog_models_artifact_repo.py:131]: I am using these eprint for debugging, will remove them before merging

File: mlflow/store/artifact/databricks_sdk_models_artifact_repo.py
Comment (harupy) [inline mlflow/store/artifact/databricks_sdk_models_artifact_repo.py:24]:
```suggestion
```

let's remove this.

File: mlflow/store/artifact/databricks_sdk_models_artifact_repo.py
Comment (harupy) [inline mlflow/store/artifact/databricks_sdk_models_artifact_repo.py:28]:
```suggestion
        catalog, schema, model = model_name_parts
```

we can use unpacking:

File: mlflow/store/artifact/databricks_sdk_models_artifact_repo.py
Comment (harupy) [inline mlflow/store/artifact/databricks_sdk_models_artifact_repo.py:62]: should we use the overwrite argument to true?

File: mlflow/store/artifact/databricks_sdk_models_artifact_repo.py
Comment (harupy) [inline mlflow/store/artifact/databricks_sdk_models_artifact_repo.py:29]:
```suggestion
        self.model_base_path = f"/Models/{catalog}/{schema}/{model}/{model_version}"
        self.client = _get_databricks_workspace_client()
```

can we instantiate the client here and reuse it?

File: mlflow/store/artifact/databricks_sdk_models_artifact_repo.py
Comment (harupy) [inline mlflow/store/artifact/databricks_sdk_models_artifact_repo.py:62]: Does `workspace_client.files.upload` support streaming upload?

Comment (shichengzhou-db) [reply mlflow/store/artifact/databricks_sdk_models_artifact_repo.py:62]: It does not. But there's plan for this near end of Q4
We're implementing this right now to unblock the small subset of SEG customers (better than having no mlflow registry at all)

Comment (shichengzhou-db) [reply mlflow/store/artifact/databricks_sdk_models_artifact_repo.py:62]: good catch, yes absolutely!

File: mlflow/store/artifact/databricks_sdk_models_artifact_repo.py
Comment (harupy) [inline mlflow/store/artifact/databricks_sdk_models_artifact_repo.py:85]: is there a way to write in chunks like this?

```
for chunk in contents:
    f.write()
```

File: mlflow/environment_variables.py
Comment (harupy) [inline mlflow/environment_variables.py:659]:
```suggestion
#: existing model artifact repo classes. mlflow
```

I don't think we should reveal the internal plan.

File: mlflow/store/artifact/databricks_sdk_models_artifact_repo.py
Comment (harupy) [inline mlflow/store/artifact/databricks_sdk_models_artifact_repo.py:42]:
```suggestion
            relative_path = posixpath.relpath(directory_entry.path, self.model_base_path)
```

File: mlflow/store/artifact/databricks_sdk_models_artifact_repo.py
Comment (serena-ruan) [reply mlflow/store/artifact/databricks_sdk_models_artifact_repo.py:29]: Is catalog/schema used anywhere else? If not can we use `"/Models/{model_name.replace(".", /)}/{model_version}"`

File: mlflow/store/artifact/databricks_sdk_models_artifact_repo.py
Comment (shichengzhou-db) [reply mlflow/store/artifact/databricks_sdk_models_artifact_repo.py:85]: good point. I took a look at databricks-sdk-py, [they're indeed using the `stream` field in the requests call](https://github.com/databricks/databricks-sdk-py/blob/79b096fb414d87ea5c8d4484687544ce63bbe266/databricks/sdk/service/files.py#L833). So we should be able to download the 5GB file in small chunks.
Your comment reminds me that the _upload_to_cloud call is also subject to similar OOM issue(I need to supply a file handle to `upload` instead of reading the file into memory and wrap it as a Bytes IO object)

File: mlflow/store/artifact/databricks_sdk_models_artifact_repo.py
Comment (shichengzhou-db) [reply mlflow/store/artifact/databricks_sdk_models_artifact_repo.py:62]: I think I misunderstood your question here. Yes it should support streaming upload (given the 5GB limit). Will fix the code here to take advantage of streaming.

File: mlflow/store/artifact/databricks_sdk_models_artifact_repo.py
Comment (harupy) [inline mlflow/store/artifact/databricks_sdk_models_artifact_repo.py:78]: We can simplify this code using assignment expression:

```python
while chunk := contents.read(DOWNLOAD_CHUNK_SIZE):
    f.write(chunk)
```

File: mlflow/store/artifact/databricks_sdk_models_artifact_repo.py
Comment (harupy) [inline mlflow/store/artifact/databricks_sdk_models_artifact_repo.py:72]: can we remove this?

File: tests/store/artifact/test_databricks_sdk_models_artifact_repo.py
Comment (harupy) [inline tests/store/artifact/test_databricks_sdk_models_artifact_repo.py:170]: We should use `monkeypatch.setenv` when we need to set an environment variable for testing purposes.

Comment (harupy) [reply tests/store/artifact/test_databricks_sdk_models_artifact_repo.py:170]:
```suggestion
def test_mlflow_use_databricks_sdk_model_artifacts_repo_for_uc(tmp_path, monkeypatch):
    monkeypatch.setenv("MLFLOW_USE_DATABRICKS_SDK_MODEL_ARTIFACTS_REPO_FOR_UC", "true")
```

File: mlflow/store/artifact/databricks_sdk_models_artifact_repo.py
Comment (harupy) [inline mlflow/store/artifact/databricks_sdk_models_artifact_repo.py:24]:
```suggestion
        self.client = _get_databricks_workspace_client()
```

can we rename this attribute?

File: mlflow/store/artifact/databricks_sdk_models_artifact_repo.py
Comment (harupy) [inline mlflow/store/artifact/databricks_sdk_models_artifact_repo.py:33]: `list_directory_contents` result is paginated; can we handle pagination in the response to ensure we retrieve all files in the directory?

File: mlflow/store/artifact/cloud_artifact_repo.py
Comment (harupy) [inline mlflow/store/artifact/cloud_artifact_repo.py:281]: can we check if `self` is an instance of `DatabricksSDKModelsArtifactRepository` instead?

File: mlflow/store/artifact/databricks_sdk_models_artifact_repo.py
Comment (shichengzhou-db) [reply mlflow/store/artifact/databricks_sdk_models_artifact_repo.py:33]: from the sdk documentations, it looks like the page token and page size are both optional, this alone should get us an iterator to exhaust the full list of directory entries

Comment (harupy) [reply mlflow/store/artifact/databricks_sdk_models_artifact_repo.py]: https://databricks-sdk-py.readthedocs.io/en/latest/workspace/files/files.html#databricks.sdk.service.files.FilesAPI.list_directory_contents says:

> We recommend not to set this value unless you are intentionally listing less than the complete directory contents. If unspecified, at most 1000 directory entries will be returned. The maximum value is 1000. Values above 1000 will be coerced to 1000.

Comment (shichengzhou-db) [reply mlflow/store/artifact/databricks_sdk_models_artifact_repo.py]: My understanding is, e.g if there's 1002 entries
The iterator will first fetch 1000 entries,
when `for directory_entry in resp:` exhaust that 1000 entries, the iterator `resp` will have to talk to the Files API to fetch the next page with the remaining 2 entries

Comment (shichengzhou-db) [reply mlflow/store/artifact/databricks_sdk_models_artifact_repo.py]: let me read their impl and get back to you, maybe my assumption was wrong (but it'd be a bit bizzare that they provide a files listing API, and if optional parameters are not provided then it return incorrect results).

Comment (shichengzhou-db) [reply mlflow/store/artifact/databricks_sdk_models_artifact_repo.py]: ah.. i think you're right
`To list all of the entries in a directory, it is necessary to continue requesting pages of entries until the response contains no next_page_token. `

Comment (shichengzhou-db) [reply mlflow/store/artifact/databricks_sdk_models_artifact_repo.py]: it's kinda weird that the API doesn't return a page_token.. it only returned an iterator..

Comment (shichengzhou-db) [reply mlflow/store/artifact/databricks_sdk_models_artifact_repo.py]: ok.. so it seems like my original assumption is correct:
https://github.com/databricks/databricks-sdk-py/blob/main/databricks/sdk/service/files.py#L896-L945
This code takes care of all the page_token logic and makes sure to exhaust all entries out of a directory.
I don't really understand why they would still want to expose the page_size and page_token there..

Comment (harupy) [reply mlflow/store/artifact/databricks_sdk_models_artifact_repo.py]: Sounds good. looks like the doc is incorrect.

**Review by harupy — APPROVED** [review]
LGTM!

**Comment by harupy** [comment]
do we need to ship this in 2.17.0? If yes, can you add the `patch-2.17.0` label?

**Comment by shichengzhou-db** [comment]
> do we need to ship this in 2.17.0? If yes, can you add the `patch-2.17.0` label?
It's okay, we can wait for next releases. Also I may have follow-up changes to auto detect SEG workspace and remove the need of user manually opting in via the environment variable

**Pull request merged**
